### ISSUE TITLE
Roboelectric support missing

### ISSUE THREAD

**Issue by vvictor10** [realm/realm-java#904]
Referring to: https://github.com/robolectric/robolectric/issues/1389

Can we please get an update on this issue. We have tons of integration test we need to run outside of an emulator/device and roboelectric is the only way for us to shadow the Android classes. Without the required support on Realm, we are pretty much blocked before we make the move over from DB4O to Realm.

The other two major sticking points are support for null and the lack of a high level API list for performing migrations.

Thanks a lot for the support.

**Comment by bmunkholm** [comment]
It is for sure on the list and being prioritized. It's not actively being worked on at the moment since other features are simply prioritized above. 
The two things that are at the top of the list is Null support (being worked on in the core database) and migrations which you can follow here: #880. So stay tuned for those! For both those features there are workarounds though. 

We are in process of moving all our tasks in our private Asana list to github issues for full transparency of our priorities. Now the issue is here, people are more than welcome to +1 this issue.

**Comment by vvictor10** [comment]
Thanks for the update. I forgot to mention the 'cascade-delete' feature. Is that anywhere near the top?

**Comment by SidhNor** [comment]
+1

**Comment by kaushikmukunda** [comment]
+1

**Comment by cchilton** [comment]
+1

**Comment by mustafa01ali** [comment]
+1

**Comment by mmddimov** [comment]
+1

**Comment by chandragm** [comment]
+1

**Comment by SidhNor** [comment]
+1

**Comment by mmvicsonvictor** [comment]
It was mentioned on the monthly call this week that you guys would be sharing your roadmap, which should give us an idea of whats coming in future releases. There are some blocking issues and would like to know, if possible, when they would possibly be addressed. Any idea when and how that roadmap is going to be shared.

**Comment by caseycrites** [comment]
+1

**Comment by IGZjuantrias** [comment]
+1

**Comment by RobertoArtiles** [comment]
+1 Really important to make it happen. And I wouldn't completely rely on Robolectric. It's obviously more important for you than for them. Currently it's a pretty serious blocker for a lot of companies to switch to Realm.

**Comment by mmvicsonvictor** [comment]
I missed the meetup on Tuesday, can you guys please share the roadmap which was supposed to be presented during the meetup. There are some major bottlenecks with Realm, which is stopping us from switching over and I want to get a sense of where they are on the priority list and if possible a timeline on when we can see them addressed. Thanks.

**Comment by alexustinovsm** [comment]
+1

**Comment by aradetegen** [comment]
+1

**Comment by baolongnt** [comment]
+1

**Comment by JackWink** [comment]
+1

**Comment by mmvicsonvictor** [comment]
Can we please get an update on this issue? Thanks.

**Comment by timanglade** [comment]
Sorry we don’t have much new to announce here… This is still low on our priority list at the moment as we are dealing with a lot of support, bugfixes and major blockers such as passing objects across threads, null support, more POJO-like models, etc. We’ll be sure to add more relevant details to this thread as soon as there is anything new we can say.

**Comment by Jberlinsky** [comment]
+1. This would be a pretty easy win.

**Comment by wryday** [comment]
+1 Would love to use Realm and Robolectric

**Comment by davidwilemski** [comment]
+1

**Comment by balanon** [comment]
Ah please bump this in priority! This would really be spectacularly helpful. 

+1

**Comment by moosingin3space** [comment]
+1 unit testing is important, and to be truly effective must be performant.

**Comment by mmvicsonvictor** [comment]
@timanglade Thanks for responding to my query. I would love to see all the ones you mentioned addressed sooner rather than later. But you can clearly see how much people want to see Robolectric support for Realm from all the +1s this thread has received. There was a mention of the roadmap/priority list being shared with the community, but I have not seen that happen. Is it possibly for you guys to share that information?

**Comment by peplin** [comment]
Since shipping native support for any of the arbitrary platforms someone might be running might be too much to ask, is it more feasible that you could ship support for the platforms used by popular cloud CI services (e.g. Travis CI runs on 64-bit Linux)?

**Comment by mmvicsonvictor** [comment]
Agree with @peplin The most popular platforms is something we should be able to gather and I would assume bamboo on linux would be on that list.

**Comment by timanglade** [comment]
@mmvicsonvictor yes, the roadmap hasn’t been published yet because there have been a lot of discussion in the wake of our SF user meetup in March, and we’re still nailing down some changes to how we think & work towards 1.0. We’ve noticed all the new +1’s that have been added to this issue recently, which spurred the team to chat again about how to prioritize this. It’s not that we don’t care, it’s just that we only have 3 people working on realm-java and we’re already maxed out just dealing with support, bugfixes and feature requests that have more +1’s than this (yes, we are [hiring](http://realm.io/jobs/)). We’ll update this issue if we prioritize this higher, and we’ll share a public roadmap as soon as we can on our blog.

**Comment by nilsi** [comment]
Any progress on this?
Just to be clear, if I have an android app with one main activity that uses realm. It's impossible for me to test the App with Robolectric right now? Or are there any hacks that I can use to mock Realm or whatever?

**Comment by nilsi** [comment]
+1

**Comment by vvictor10** [comment]
@timanglade Haven't be following Realm closely for a few weeks now. Was a roadmap released? If yes, can you please provide a link. Thanks.

**Comment by timanglade** [comment]
We haven’t published a roadmap yet so here is an update. (You can see our priorities in the open on Github btw: all of the stuff we are aiming to complete before 1.0 is [labelled P1](https://github.com/realm/realm-java/labels/P1).) As you can see from this issue history, we bumped this up to a P1 a couple of weeks ago, and actually scheduled it pretty high — we’re only aiming to complete a few other features before we get to Roboelectric, namely: Migrations, Null support, Fine-grained Notifications and Async/Handoff (all based on the pain points users are reporting, and how many people are reporting them). The good news is that all 4 of those have PRs open so we should be starting work on Roboelectric very soon (probably as soon as one of those features get merged and one of us is free to start working on a new feature). We recently [added two more Android developers to the team](https://realm.io/news/nabil-adam-mulong-join-realm/) to try to deliver all these key features as quickly as possible, and [we are still hiring](https://realm.io/jobs/#android-developer-with-brstrong-ndk-experience) to go even faster. Like you, I wish we had added Roboelectric support months ago, but we need to continue to prioritize answering support questions & fixing bugs above any feature work, and to prioritize feature work 100% based on user feedback. (We also welcome meta-feedback on how we handle priorities and communications, so if you think there is anything we should change there, please speak up!)

**Comment by vvictor10** [comment]
Thanks a lot for that transparency, this is one of the most important reasons why we are still actively looking at Realm for our company wide stack. All the issues you mentioned are major pain points for us. 
Another one which I do not see on your list is the "cascade-delete" functionality driven by annotation options. I think that would be a huge help. But Migrations, Null support and fine grained notifications to the listeners would be super useful as well. 

And the query linking would also be super useful. (https://github.com/realm/realm-java/issues/792) I miss it real bad.

A quick question - "Async/Handoff" does that refer to the ability to pass realm objects across threads?

Thanks @timanglade Have a great weekend.

**Comment by timanglade** [comment]
> Thanks a lot for that transparency, this is one of the most important reasons why we are still actively looking at Realm for our company wide stack. All the issues you mentioned are major pain points for us.

We completely understand! We have some ways to go before 1.0 and we appreciate everybody’s excitement to see us get there. Mostly we feel bad we haven’t been able to make happy users out of everybody already :D

> Another one which I do not see on your list is the "cascade-delete" functionality driven by annotation options. I think that would be a huge help. 

Agreed, #1104 is also a P1 (i.e. delivered before 1.0), just ranked after these top 4–5 features at the moment.

> […] And the query linking would also be super useful. (#792) I miss it real bad.

That’s backlogged at the moment, so don’t hesitate to put a +1 detailing your use-case in there to help us prioritize more accurately.

> A quick question - "Async/Handoff" does that refer to the ability to pass realm objects across threads?

Sort of! The first iteration will be delivering methods that let you make a query or update/insert from the main thread, execute it on a background thread, and receive results back on the main thread. We think that will cover the wide majority of cases people need while preserving great ease of use (coherence of your data across threads) and performance. That said, there are other use cases and the the hooks underneath this first iteration will let us consider other, maybe more direct ways to pass objects across threads; for example we’re considering ways for you to manually pin/unpin objects from one thread to another.

I hope this helps! Never hesitate to ping us on Github issues if you have any Q’s or privately by email I’m tim@realm.io.

**Comment by vvictor10** [comment]
Thanks @timanglade. I know its hard to provide deadlines, but is there anyway you can give me an approx timeline when the null support and the new migration APIs would be available as part of the official release? Those two are the top most concerns right now and making us use other alternatives. I would like to get an understanding of how long we need to hold off on Realm. Thanks.

**Comment by timanglade** [comment]
Hi. Better migrations are coming but there should be nothing blocking you with the existing ones. Yes they’re a bit tricky to get right, but you can do with them anything you can do with the next revision — if you have any issues at all don’t hesitate to open an issue on the #realm tag on StackOverflow so we can help. Regarding Null, we will start adding support for this type by type (strings & binaries) in the next 2–3 weeks. I hope that helps!

**Comment by Renascienza** [comment]
+1

**Comment by upelsin** [comment]
+!

**Comment by McPo** [comment]
+1
Greatly appreciate the recent detailed update

**Comment by baolongnt** [comment]
@timanglade  Looks like priority for this feature has been downgraded from P1 to P2. Is there any context you can provide us?

**Comment by cmelchior** [comment]
@baolongnt It is  because we redefined what P1 ment. P1 was changed from meaning "must have to make a 1.0 release" to mean "No further API breaking anticipated with no outstanding bugs". Robolectric doesn't really fit P1 with the new meaning. We are tracking the relative order between items in P1/P2 in another tool and there Robolectric hasn't changed position. So this change is mostly an internal project management thing.

**Comment by baolongnt** [comment]
Thanks for the clarification. I am happy to hear that the feature is still in the top of the queue.

**Comment by dmitry-zaitsev** [comment]
+1

**Comment by cjkrieseSF** [comment]
:+1:

**Comment by kishisaka-basis** [comment]
+ 1

**Comment by YukiMatsumura** [comment]
+1

**Comment by yss44** [comment]
+1

**Comment by ko3ak1990** [comment]
+1

**Comment by fwouts** [comment]
+1

**Comment by guillermomuntaner** [comment]
+1

**Comment by enterline** [comment]
We are currently using Robolectric and our solution so far has been to wrap Realm and mocking some things like `RealmResults`, `RealmQuery` etc. 

The latest 0.84.1 release has made classes like `RealmResults` final. Not sure what the rationale behind making these classes final was, but it makes it tremendously more difficult to test and mock. 

Until we get full Robolectric support, is there any way that you can remove the `final` from these classes?

**Comment by cmelchior** [comment]
@enterline Sorry for the late response. You can use PowerMock to the around the final classes: https://code.google.com/p/powermock/wiki/MockitoUsage13

They have been made final so people don't accidentally subclass them as their internals are quite complex.

**Comment by cmelchior** [comment]
Plan for approaching this would be something like this:

- [ ] 1. Look into how Robolectric loads native resources (Gradle plugin / java.library.path) judging by  https://github.com/robolectric/robolectric/issues/1389
- [ ] 2. We already have Realm distributed for Mac OS X. We should use that to test integration with Robolectric.
- [ ] 3. Once we have Mac OS X working we need to determine how/which platforms we wish to support. It will most likely require us upgrading our build pipeline to release binaries for other platforms as well: Windows (DLL), Linux (Which variants).

**Comment by enterline** [comment]
Thanks for the response @cmelchior . In our case we have quite a lot of tests and  the interactions with PowerMock, Robolectric and Jacoco do not play nice with each other. 

For simple tests, it can work but we lose code coverage results with Jacoco. For more complex tests, it just doesn't quite work. This is more of a PowerMock, Robolectric and Jacoco issue.

Right now I am exploring wrapping some of these final classes so we can upgrade to 0.84.1 and test and mock a bit easier. We already wrap Realm itself so this would just be an extension of this pattern.

Anyway, thanks for the help and we are looking forward to full Robolectric support!

**Comment by Hazealign** [comment]
+1

**Comment by alexgarciar** [comment]
+1

**Comment by anitas3791** [comment]
+1

**Comment by andriipanasiuk** [comment]
+1

**Comment by scijoker** [comment]
+1

**Comment by JAertgeerts** [comment]
+1

**Comment by renannery** [comment]
+1

**Comment by PhilippeBoisney** [comment]
+1000

**Comment by MonStar1** [comment]
+1

**Comment by alison-feder-cooper** [comment]
+1

**Comment by poldz123** [comment]
+1

**Comment by alekseichuk** [comment]
+1

**Comment by robsn** [comment]
+1

**Comment by Zhuinden** [comment]
If [Realm supports Windows now as a platform](https://realm.io/news/realm-microsoft-xamarin-1-0-azure-windows-desktop/?utm_source=newslist&utm_campaign=xamarin-launch&utm_medium=email), does that mean it's possible to refactor the notification system so that Realm-Java will soon be able to run on JVM and unit tests?

**Comment by dalinaum** [comment]
@Zhuinden There is another issue. We cannot use Transformer API for this issue, so we should find a work-around or implement something.

**Comment by groys** [comment]
+1

**Comment by xian** [comment]
@cmelchior -- Robolectric maintainer here, anything we can do to help with this?

**Comment by ajaysaini-sgvu** [comment]
+1

**Comment by cmelchior** [comment]
@xian Thanks for checking in. At this point, I think it is mostly on us. We need to free some cycles to move this closer to completion. Biggest blocker is us modifying our CI setup so we can build and release separate libraries with the native code needed for the different platforms: Mac, Linux, Windows.

**Comment by virl** [comment]
Supporting Robolectric may require fixing Looper support in Robolectric, because, as I understand, Realm heavily relies on them for cross-thread object exchange:
https://github.com/robolectric/robolectric/issues/2977
https://github.com/robolectric/robolectric/issues/1993

**Comment by josias1991** [comment]
👍

**Comment by venugv** [comment]
Any update on the timeline for this?  Thanks folks for taking this up 👍

**Comment by The0ldM0nk** [comment]
:+1:

**Comment by danielchristopher1** [comment]
+1

**Comment by subaru44k** [comment]
+1

**Comment by CNlichao** [comment]
+1

**Comment by poldz123** [comment]
+1

**Comment by lucaspommateau** [comment]
+1

**Comment by filipebrandao** [comment]
+1

**Comment by tfabio** [comment]
+1

**Comment by Zhuinden** [comment]
The project is open-source. This is something that anyone could add.

**Comment by virl** [comment]
@Zhuinden No, project is not open source, because it exists only due/for financial support of single company.

And no, it is not anyone could add, because implementing such feature even in Realm sources is huge investment that most users don't have time/money for.

And  even if they did, it requires forking and global refactoring of Robolectric, because Realm cannot work with Robolectric due to its broken async testing paradigm.

And even all of this is done by third party, it will not be continuously supported and patched to recent Realm releases without support from Realm company.

**Comment by The0ldM0nk** [comment]
IMHO since robolectric's maintainer has [already offered support](https://github.com/realm/realm-java/issues/904#issuecomment-284549524) and for all the feasible reasons involved, _Realm maintainers should find a way to work it out together with robolectric team_ for Realm to maintain it's competitiveness with alternative options.

Else, provide the community a solid way to test the sh!te out of realm (using other widely frameworks?) otherwise I don't see a point where one would want to use realm for production scenarios.

**Comment by virl** [comment]
@The0ldM0nk Yes, but really what is needed is Robolectric to fix their broken async testing paradigm and allow full-fledged Loopers to exist in tests. And also implement analogue to XCTestExpectation.

In other words, it's not something specific to the Realm, it's the same problem of Robolectric with ANY OTHER async code. It's just Realm is very popular example of such library.

**Comment by lordplagus02** [comment]
+1

**Comment by vciancio** [comment]
+1

**Comment by nongdenchet** [comment]
Hi guys any update on this?

**Comment by xian** [comment]
> @The0ldM0nk Yes, but really what is needed is Robolectric to fix their broken async testing paradigm and allow full-fledged Loopers to exist in tests. And also implement analogue to XCTestExpectation.
> 
> In other words, it's not something specific to the Realm, it's the same problem of Robolectric with ANY OTHER async code. It's just Realm is very popular example of such library.

Robolectric's async/Looper model is significantly revised and improved in 4.3. I think the main blocker now is availability of native libs.
